Wikipedia/Code/SearchViewController.swift

```diff
@@ -557,18 +557,39 @@ extension SearchViewController: CollectionViewHeaderDelegate {
 }
 
 extension SearchViewController: ArticleCollectionViewControllerDelegate {
-    func articleCollectionViewController(_ articleCollectionViewController: ArticleCollectionViewController, didSelectArticleWithURL: URL) {
+    func articleCollectionViewController(_ articleCollectionViewController: ArticleCollectionViewController, didSelectArticleWithURL: URL, at indexPath: IndexPath) {
+        funnel.logSearchResultTap(at: indexPath.item)
         saveLastSearch()
-        funnel.logSearchResultTap()
     }
 }
 
 extension SearchViewController: SearchLanguagesBarViewControllerDelegate {
     func searchLanguagesBarViewController(_ controller: SearchLanguagesBarViewController, didChangeCurrentlySelectedSearchLanguage language: MWKLanguageLink) {
+        funnel.logSearchLangSwitch(source)
         search()
     }
 }
 
+extension SearchViewController {
+    private var source: String {
+        guard let navigationController = navigationController, navigationController.viewControllers.count > 0 else {
+            return "unknown"
+        }
+        let viewControllers = navigationController.viewControllers
+        let viewControllersCount = viewControllers.count
+        if viewControllersCount == 1 {
+            return "search_tab"
+        }
+        let penultimateViewController = viewControllers[viewControllersCount - 2]
+        if penultimateViewController is WMFArticleViewController {
```

Comment:
Might be nice to decouple this as a protocol - `AnalyticsSearchSourceProviding` so this doesn't have to check specific classes. Not a necessary fix, just an optional enhancement